### PR TITLE
Fixes #2538

### LINKED ISSUE #2538: Recipient headers (To, CC, BCC) can exceed maximum line length
It seems that for long CC most servers do not accept messages created by K9. 

I've found two bug reports that probably pertain to the same problem - both closed without solution:
https://github.com/k9mail/k-9/issues/2153 , https://github.com/k9mail/k-9/issues/2090

Gmail returns:
209.85.233.26 failed after I sent the message.
Remote host said: 550-5.7.1 [212.77.101.11      11] Our system has detected that this message is
550-5.7.1 not RFC 5322 compliant:
550-5.7.1 'From' header is missing.

Yahoo:
98.136.216.25 failed after I sent the message.
Remote host said: 554 Message not allowed - [299]

Below is a sample header taken from mailer daemon reply. In addresses I've replaced small letters and numbers with *.
````
Return-Path: <********@**.**>
Received: (wp-smtpd smtp.wp.pl 21731 invoked from network); 9 May 2017 12:39:35 +0200
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=wp.pl; s=1024a;
          t=1494326375; bh=GxJNyNAIze0/sLcmzs809CUhriVEk0kSCgvT6Q6V+pY=;
          h=Subject:To:CC:From;
          b=ch4amAgjxkXn+UESSOnoSNjZZoFrKztgbsFVUhGx0RtrtCzePS4z124n+tBz3sgP9
           iBIlARMAKDTf7y4a3eZjBzkB/FDC2X0lPn169X1x9NWvGZwAxjTBJBVPWdVajlSD+D
           j0Ruc3Gug6pCPDfXGT+WIxpP+G6AR/VTrYjJETl4=
Received: from ******-**********.*********.** (HELO [**.***.***.*]) (********@**.**@[**.***.***.***])
          (envelope-sender <********@**.**>)
          by smtp.wp.pl (WP-SMTPD) with ECDHE-RSA-AES256-GCM-SHA384 encrypted SMTP
          for <*******@****************.**>; 9 May 2017 12:39:35 +0200
Date: Tue, 09 May 2017 12:39:23 +0200
User-Agent: K-9 Mail for Android
In-Reply-To: <********-****-****-****-************@****************.**>
References: <**************.********@**.**> <********-****-****-****-************@****************.**>
MIME-Version: 1.0
Content-Type: text/plain;
 charset=utf-8
Content-Transfer-Encoding: quoted-printable
Subject: =?UTF-8?Q?R*=3A_4*_-_20_***_*=C3=B3=C5=BA****_-?= =?UTF-8?Q?_****_***_*****_**=C4=99***=3A-=29?=
To: A***** <*******@****************.**>,A********* R******** <****-*****@**.**>
CC: Z***** C********** <******.***********@*****.***>,P**** G******** <*.*********@******.***>,=?UTF-8?Q?T*****_=C5=81****?= <******.****@**.**>,J**** M******* <*.********@*********.***.**>,=?UTF-8?Q?D******_=C5=BB*******?= <***@*******.***>,I**** W*********** <*****@**************.**>,******** <********@******.**>,"********.*******" <********.*******@******.**>,****** <******@****.**>,"*****.*****" <*****.*****@*******.***>,******* <*******@******.****.**>,"*.******" <*.******@***.********.**>,**** <****@*****.***.**>,************** <**************@****.**>,M***** G**** <******.*****.****@*****.***>,"******.**********" <******.**********@*****.***>,"********.*******" <********.*******@*****.***>,******** <********@*****.***>,*** <***@******.**>,"*.******" <*.******@*******.**>,***** <*****@**************.***.**>,************** <**************@*****.***>,**** <****@**.**>,************** <**************@*****.***>,********** <**********@*****.***>,"*.*********"
<*.*********@*****.***>
From: K******** G**** <********@**.**>
Message-ID: <********-****-****-****-************@**.**>
X-WP-MailID: ********************************
X-WP-AV: skaner antywirusowy Poczty Wirtualnej Polski
X-WP-SPAM: NO 000000A [ceOE]    
````
### Environment
K-9 Mail version:
5.206

Android version:
6.0.1

Account type (IMAP, POP3, WebDAV/Exchange):
IMAP

### PR DESCRIPTION
**Pull request by blakawk**
- Implemented MimeUtility.foldAndEncode according to RFC 2822
   - In case string exceeds 78 characters, split it in two
   - Identified drawbacks:
     - replace horizontal tabs with space as well as contiguous whitespace with only one space
     - header name is not taken into account in 78 characters limit

Please ensure that your pull request meets the following requirements - thanks !

* Follows our existing [codestyle](https://github.com/k9mail/k-9/wiki/CodeStyle).
* Does not break any unit tests.
* Contains a reference to the issue that it fixes.

### CODE REVIEW

**Comment by philipwhiuk** [comment]
This needs some new tests before it can be merged.

**Comment by blakawk** [comment]
@philipwhiuk I added tests for that (testing long subject and CC:, over 998 characters), is that enough ? Thanks !

**Comment by philipwhiuk** [comment]
I kind of expected you to just directly test foldAndEncode. The tests you've added are good though.

**Comment by blakawk** [comment]
@philipwhiuk is this what you were expecting ?

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java [hunk 1]
```diff
@@ -83,7 +83,16 @@ public final void parse(InputStream in) throws IOException, MessagingException {
...
+    public final void parse(InputStream in, boolean recurse) throws IOException, MessagingException {
+        MimeConfig parserConfig  = new MimeConfig();
+        parserConfig.setMaxHeaderLen(-1); // The default is a mere 10k
+        parserConfig.setMaxLineLen(-1); // The default is 1000 characters. Some MUAs generate
+        // REALLY long References: headers
+        parserConfig.setMaxHeaderCount(-1); // Disable the check for header count.
+        parse(in, recurse, parserConfig);
+    }
+
+    public void parse(InputStream in, boolean recurse, MimeConfig parserConfig) throws IOException, MessagingException {
```
Comment (Valodim) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java]: If we move `MessageTest` to the internet package, these can be package-local. Also please annotate `@VisibleForTesting`

File: build.gradle [hunk 2]
```diff
@@ -4,7 +4,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.0'
+        classpath 'com.android.tools.build:gradle:2.3.3'
```
Comment (Valodim) [inline build.gradle]: please don't do this in an unrelated commit

Comment (blakawk) [inline build.gradle]: Fixed it, sorry !

**Comment by cketti** [comment]
Folding rules are header-specific. Blindly folding every header is not the correct solution, especially for `To`, `Cc`, and `Bcc`. I'll write down what needs to be done in issue #2538.
